### DISCUSSION TITLE
Realm accessed from incorrect thread

### DISCUSSION THREAD

**Discussion by dharamhbtik** [realm/realm-dotnet#3339] [Q&A] [answered]
I am trying to refresh realm through background but it is giving exception.
here is my code :
```
public static void main(){

//StartRefreshingData is called from here without any thread or async call
}
public class DataManager
{
Realms.Realm _realm;
public DataManager(Func<Realm> realmDel){
_realm = realmDel();
}
private void StartRefrshingData(){
var  _worker = new Thread(RefreshData) { IsBackground = true };
 _worker.Start();
}

1 private async void RefreshData()
2 {
3           using (var realm= Realms.Realm.GetInstance(_realm.Config))
4            {
5                   _ =  realm.Refresh(); // explicit refresh only!
6                   var persons = realm.All<Person>();
7           }
8 }
}
```
When I am executing above code, it is always throwing exception 

Line 5 is giving below error : 
```
[0:] 14:33:31.388 Exception Realms.Exceptions.RealmException: Realm accessed from incorrect thread.
  at Realms.NativeException.ThrowIfNecessary () [0x00009] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Native\NativeException.cs:77 
  at Realms.SharedRealmHandle.Refresh () [0x00008] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Handles\SharedRealmHandle.cs:554 
  at Realms.Realm.Refresh () [0x00006] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Realm.cs:979 
  at  PersonManager.DataManager.RefreshData() [0x000a4] in.....
```

Line 6 is throwing below error : 
```
[0:] 14:14:43.430 Exception Realms.Exceptions.RealmException: Realm accessed from incorrect thread.
  at Realms.NativeException.ThrowIfNecessary () [0x00009] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Native\NativeException.cs:77 
  at Realms.SharedRealmHandle.CreateResults (Realms.Native.TableKey tableKey) [0x0000f] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Handles\SharedRealmHandle.cs:710 
  at Realms.RealmResults`1[T]..ctor (Realms.Realm realm, Realms.Metadata metadata) [0x00000] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Linq\RealmResults.cs:56 
  at Realms.Realm.All[T] () [0x00059] in D:\a\realm-dotnet\realm-dotnet\Realm\Realm\Realm.cs:1026 
  at PersonManager.DataManager.RefreshData() [0x000c6] in ...... 

```
As per the documentation, from https://www.mongodb.com/docs/realm-sdks/dotnet/latest/reference/Realms.Realm.html#Realms_Realm_Refresh we have to call the Refresh method in background thread 
_Remarks
On worker threads (where the SynchronizationContext) is null, this will call the blocking [Refresh()](https://www.mongodb.com/docs/realm-sdks/dotnet/latest/reference/Realms.Realm.html#Realms_Realm_Refresh) method instead. On the main thread (or other threads that have SynchronizationContext), this will wait until the instance automatically updates to resolve the task. Note that you must keep a reference to the Realm until the returned task is resolved._

Any way to prevent this incorrect thread exception?

**Comment by nirinchev** [comment]
Interesting, I have 3 comments here:
1. The code you posted seems correct to me, so I don't have an obvious explanation for why you would get this exception.
1. Why is your `RefreshData` method `async void` if it's not using any `await`s? Is it a leftover or did you remove some actual async code in there?
2. Why are you calling `Refresh` on a background thread? What are you trying to achieve?

**Reply by dharamhbtik** [reply]
2. Yes, I removed some code which actually do some other operation and are async calls.
3. This `Refresh ` method runs every 3 seconds and checks if there is any update in the localDb then sends it over API.

**Reply by nirinchev** [reply, marked as answer]
Okay, so the problem is likely related to the async calls - what happens is roughly this:

```csharp
async Task RefreshData()
{
    // Thread A - we open the Realm
    using var realm = Realm.GetInstance(_realm.Config);
    _ =  realm.Refresh(); // explicit refresh only!
    var persons = realm.All<Person>();

    await DoSomethingAsync();

    // Continuation after the await is scheduled on Thread B

    realm.Refresh(); // Accessing realm on Thread B
}
```

If you want to do check for Realm changes in the background, it's better to install a synchronization context and rely on notifications. Something like (relies on [Nito.AsyncEx.Context](https://www.nuget.org/packages/Nito.AsyncEx.Context)):

```csharp
private void StartRefrshingData()
{
    var cts = new CancellationTokenSource();
    
    // you may want to return the cancellation token source or expose a different way to
    // cancel the background processing.
    Task.Run(() => RefreshData(cts.Token));
}


private void RefreshData(CancellationToken token)
{
    AsyncContext.Run(async () =>
    {
        using var realm = Realm.GetInstance(_realm.Config);

        using var notificationToken = realm.All<Person>().SubscribeForNotifications((sender, changes) =>
        {
            // Do some processing of the changes
            var json = payload.ToJson();
            
             // We don't await the API request. This is business needs specific - if you need
             // to ensure the previous request completed before sending a new one you could 
             // force the async method to execute synchronously by calling `.Wait()`.
            _ = NotifyAPI(json);
        });

        // Wait indefinitely or until cancellation is requested
        var tcs = new TaskCompletionSource();
        token.Register(() => tcs.TrySetResult());

        await tcs.Task;
    });
}

private async Task NotifyAPI(string json)
{
    // Send the json to your API.
}
```

Note that with the proposed approach you don't need to continuously refresh the background Realm as we've now installed a SynchronizationContext (by wrapping our code into an `AsyncContext.Run`) which means Realm will notify you as soon as changes are made. If you want to bundle those changes together rather than sending them 1 by 1, you could introduce a simple throttling mechanism.
